### ISSUE TITLE
Stack overflow on 1.10 preview3

### ISSUE THREAD

**Issue by tornatae** [signal18/replication-manager#177]
Got stuck in a stack overflow repeat on virtual CentOS 7 machine. I've gotten segfaults from the stable version, hence trying the preview. Didn't record the errors for the segfaults but will if I run into them again.

Edit: I got a segfault starting with a fresh CentOS 7 Linode instance. Realize pasting  the entire logs in a comment is kind of obnoxious so here it is in a gist:

https://gist.github.com/Swashy/d9f37ce33396db140db6d468e04f6c20

**Comment by svaroqui** [comment]
Hello , i fixed that in the last commit , 

This happen with positional replication, replication-manager was created for MariaDB GTID 
Was not well tested on positional replication until last week  

- In new commit the crash does not happen and topology can be MMS 
- In new commit we should not enter this state as positional should work on MySQL at least 

Let me know about : 
- What type of package you would like ?
- What Database release using, will run some regtest based on your setup 

Thanks 
/stephane

**Comment by tornatae** [comment]
Hey thanks for the speed reply!

>What type of package you would like ?

Not following you here, could you clarify?

>What Database release using, will run some regtest based on your setup

`rpm --query centos-release`
`centos-release-7-3.1611.el7.centos.x86_64`

Using the latest version stable of MariaDB from their repos:

`mysql -V`
`mysql  Ver 15.1 Distrib 10.2.7-MariaDB, for Linux (x86_64) using readline 5.1`

I also got a segfault with a fresh centos install(just to make sure I wasn't the problem) and replication-manager install (latest preview). Literally just ran an update, installed epel-release and nginx. i've attached it at the bottom of the gist I provided.

I'm also having systemd auto-restart on failure, but hadn't experienced a loop like that prior to this. Was running it for a few days no problems.

**Edit:** My topology is one master and two slaves, here's a snippet from the configuration:

`autorejoin = true`
`autorejoin-semisync = true`
`autorejoin-flashback = true`
`autorejoin-mysqldump = true`
`mariadb-binary-path = "/usr/bin"`
`#interactive = false`

`readonly = true`
`failover-event-scheduler = false`
`failover-event-status = false`

`#multi-tier-slave=true`

**Comment by svaroqui** [comment]
Re,
Re,


I will test your setting tomorrow and send you a release that fixe your case  . 

Can you confirm you are not currently using GTID in your replication ? 
Send us a show slave status; 

If you activate GTID replication  you should be able to go 

stop slave ; 
change master to master_use_gtid=slave_pos;
start slave;
 
Stéphane Varoqui, Senior Consultant
Phone: +33 695-926-401, skype: svaroqui
http://www.mariadb.com <http://www.mariadb.com/>

**Comment by tornatae** [comment]
I would greatly appreciate the release, thanks! I am indeed using GTID replication, **NOT** positional replication. Here's a slave status, as well as a snippet from the mysql configuration:

https://gist.github.com/Swashy/1e0b3ce29da9016203b47b2c405cb2cd

**Comment by svaroqui** [comment]
Config looks very good, Can i get you mail ? i'll send you the new release.

**Comment by svaroqui** [comment]
I'm currently working on Q&A via Docker , will get soon a json of all test case per setting somewhere , that will greatly help to qualify the quality of Pre GA release

Thanks for trying it !

**Comment by svaroqui** [comment]
A we are approaching GA this new package get pure client server secure API 
the http server need to be enable in config file as default is disable because the http server does not use the API yet . 

To get the terminal  client replication-manager client is the way to go !  or protecting the http server via proxy

**Comment by svaroqui** [comment]
you can mail me at stephane@mariadb.com i will reply if you don"t wan't email in git

**Comment by tornatae** [comment]
Thanks for the new binary. Did lots of testing with it and after failing over a few times I'm still seeing issues. Sometimes rejoined slaves don't have master set properly when failing back and forth, and sometimes they need a "start slave". **Edit:** And sometimes failing over/starting an old master results in a panic! 

Here's my process for testing failover:

-Shut down mariadb on master
-Failover to a slave (works)
-Insert some test data into new master
-bring up mariadb on *old* master
-Replication manager usually fails to mysldump or rejoin via GTID.
-Fail back over
-"panic: runtime error: invalid memory address or nil pointer dereference"

Logs here: 

https://gist.github.com/Swashy/dcd2ae72de8ff8984336aaa7a5c77d1f

Full config here:

https://gist.github.com/Swashy/1529692f45bffc4fb8d25664baf05272

**Comment by tornatae** [comment]
I had not re-installed mariadb and mariadb tools on the server with replication-manager, so that's why the mysqldump rejoin was failing. GTID rejoin and other problems still stands and I'm doing more testing.

**Comment by svaroqui** [comment]
Swashy,

Thanks ! 

There is a mysqlbinlog tool provided with the package but in your case it was pointing to non exiting binary , hope you installed mariadb 10.2 as this is the one needed for remote binlog copy

**Comment by svaroqui** [comment]
Hello i have verified the issue with test case 
I think i get it fixed .

1455:		"name": "testFailoverAssyncAutoRejoinFlashback",
1456:		"result": "FAIL"
1457-	}
1458-]
--
--
1460-[
1461-	{
1462:		"name": "testFailoverSemisyncAutoRejoinFlashback",
1463:		"result": "FAIL"
1464-	}

Will come back when more testing is done

**Comment by tornatae** [comment]
Awesome!

Also here's an example of binlog rejoin and GTID reloin failing, but it's confusing because we aren't using position replication, we're using GTID:

https://gist.github.com/Swashy/8edd774044f1fa285180f3f4e4765546

In this log snippet replication-manager is attempting to restore from a binlog position that doesn't exist on the server (mysql-bin.000005) and it also finds an "empty" gtid sequence. Sometimes GTID rejoin works, but sometimes it fails like this. I'd appreciate it if you could point out what I may be doing wrong. :)

**Comment by tanji** [comment]
Have you enabled semi-sync?

**Comment by tanji** [comment]
@svaroqui when mysqlbinlog fails it doesn't return the error so we don't know why it failed in this case (some wrong parameters?)

**Comment by svaroqui** [comment]
Or it's not 
-10.2 mysqlbinlog 
- good disk located
- not row base replication

**Comment by tornatae** [comment]
@tanji I do indeed have semi-sync enabled. I'm pretty sure Mysqlbinlog is failing because the binlog file it's attempting to use doesn't exist.

FWIW I'm using mysqlbinlog that was included with mariadb-server, and here's the output:
`mysqlbinlog Ver 3.3 for Linux at x86_64`

**Comment by svaroqui** [comment]
what is output of /usr/bin/mysqlbinlog --read-from-remote-server --raw --stop-never-slave-server-id=10000 --user=replication --password= --host=db02.dbserver --port=3306 --result-file=/var/lib/rep
lication-manager/Default-server2- --start-position=4322 mysql-bin.000005

**Comment by tornatae** [comment]
Simulated a new failover and failback, here's the logs starting from the switchover. I inserted a few records into the new master and then started mariadb again on the old master:

https://gist.github.com/Swashy/f04e6b933c8b2c1b6a13fad3b54f88e6

Then I ran the command you provided on the replication manager server (separate from the cluster).

1. Ran command, no stdout output.
2. Vim and strings view of the ouput file, `/var/lib/replication-manager/Default-server2-mysql-bin.000005`:
 https://gist.github.com/Swashy/b18601f13cdd3752e713e67aa636df05

Those are indeed the test values I inserted into the database.
